### PR TITLE
[ROCm][Perf] Use wvSplitK for single-output GEMMs

### LINKED ISSUE #52631: [Performance][ROCm] shared_expert_gate ([1, K] weight) misses every skinny-GEMM path and costs 13.7% of the decode step
## Summary

On ROCm, `rocm_unquantized_gemm_impl` dispatches on `m = weight.shape[0]`. Its two skinny-GEMM branches cover `m > 8` (wvSplitK) and `m % 4 == 0` (LLMM1), so for `m <= 8` only `m ∈ {4, 8}` are covered — **`m ∈ {1, 2, 3, 5, 6, 7}` falls through** to `torch.nn.functional.linear` → rocBLAS. (The `n` conditions also differ: wvSplitK takes `0 < n <= 5`, LLMM1 only `n == 1`.)

`m == 1` is not a rare shape. In transformers today, `self.shared_expert_gate = torch.nn.Linear(config.hidden_size, 1, bias=False)` — i.e. a `[1, hidden]` weight — appears in **`qwen2_moe`, `qwen3_next`, and `qwen3_5_moe`**, invoked once per layer per token to produce the scalar that gates the shared expert's contribution. (`qwen3_moe` and `qwen3_vl_moe` have no shared expert, so they are unaffected; `glm4_moe` has a shared expert but no gate.) Because `bias=False` there, `bias is None` always holds for this call.

For a `1×2048 @ 2048×1` product, rocBLAS selects `Cijk_Ailk_Bljk_BBS_BH_MT128x32x16_..._GSU1_...` — a **128×32 macro-tile with no split-K**. That is one workgroup on a 40-CU GPU serially walking `k=2048`: **~140 µs to move 4 KiB of weights**, against **~7.9 µs** for every width the skinny ladder does cover.

Measured on a Qwen3.5-MoE 35B-A3B (40 layers, hidden 2048, 256 experts, 8 active), batch-1 decode, gfx1151:

| | |
|---|---|
| cost per call | 137 µs |
| calls per decode step | 40 (one per layer) |
| **total** | **5.48 ms of a 40.06 ms step = 13.7%** |
| decode after fixing it | 40.06 → 34.79 ms/step, **24.53 → 28.32 tok/s (+15.5%)** |

## Environment

- gfx1151 (Radeon 8060S / Strix Halo), ROCm 7.14, torch 2.11.0+rocm7.14.0
- vLLM 0.27.1 (the same dispatch conditions are present on `main` today)
- `VLLM_ROCM_USE_SKINNY_GEMM` at its default (`True`)

## Where

`vllm/model_executor/layers/utils.py`, in `rocm_unquantized_gemm_impl`:

```python
    if use_skinny:
        x_view = x.reshape(-1, x.size(-1)).contiguous()
        if m > 8 and 0 < n <= 5:                                     # m == 1 fails
            ...
        elif m % 4 == 0 and n == 1 and k <= 8192 and bias is None:   # 1 % 4 != 0, fails
            ...

    if rocm_aiter_ops.is_tgemm_enabled():
        ...
    return torch.nn.functional.linear(x, weight, bias)   # <-- m == 1 lands here
```

Note `is_tgemm_enabled()` requires `on_gfx950()`, and `use_aiter_triton_gemm()` matches only a hardcoded shape whitelist, so on gfx11xx there is no other path to catch this.

## Reproducer

Standalone, no checkpoint needed. Part A needs only torch; Part B additionally needs vLLM: https://gist.github.com/wenis/ea2a42350797771304a2d8ea8c7848f7

Output on the environment above, **stock vLLM**, `--k 2048 --iters 200`:

```
device : AMD Radeon 8060S Graphics
torch  : 2.11.0+rocm7.14.0
shape  : x[1, 2048] @ w[m, 2048].T  (batch-1 decode), bfloat16

=== Part A: torch.nn.functional.linear (vLLM's fallback path) ===
   m    us/call     GiB/s   note
   1     141.17      0.03   <- shared_expert_gate; 4 KiB of weights
   2      37.45      0.20
   3      40.33      0.28
   4      37.82      0.40
   8      39.65      0.77
  12      38.00      1.20
  16      39.09      1.56
  32      40.12      3.04
  64      37.81      6.46

m=1 as a reduction (x * w).sum(-1) :     5.47 us
m=1 via F.linear                   :   141.17 us   (25.8x slower)

=== Part B: vLLM's rocm_unquantized_gemm dispatch ===
   m    us/call   expected path
   1     139.93   *** NO SKINNY PATH -> rocBLAS fallback ***
   2      37.16   *** NO SKINNY PATH -> rocBLAS fallback ***
   3      38.82   *** NO SKINNY PATH -> rocBLAS fallback ***
   4       7.90   LLMM1 (m % 4 == 0, n == 1)
   8       7.88   LLMM1 (m % 4 == 0, n == 1)
  12       7.99   wvSplitK (m > 8, n <= 5)
  16       7.95   wvSplitK (m > 8, n <= 5)
  32       8.01   wvSplitK (m > 8, n <= 5)
  64       7.96   wvSplitK (m > 8, n <= 5)
```

Two separate things show up here, and I think only the first is vLLM's to fix:

1. **The dispatch gap.** `m = 1` costs **139.93 µs** against **~7.9 µs** for every covered width — **17.7x** — while moving *less* data than any of them.
2. **rocBLAS is additionally pathological at `m == 1` specifically.** In Part A, which never touches vLLM's ladder, `m = 1` costs **141 µs** while `m ∈ {2, 3, 4, 8, ..., 64}` all sit at **37–40 µs**. So `m = 2` and `m = 3` miss the skinny ladder too, but cost ~3.7x less than `m = 1` does. The degenerate single-row output is its own kernel-selection problem, independent of vLLM.

## Attribution

The 137 µs in the table above was measured on the full model and then traced, not inferred from reading the guard:

1. vLLM's own torch profiler (`--profiler-config.profiler=torch`), 51 decode steps under representative serving flags with CUDA graphs **on**. Kernel `Cijk_..._MT128x32x16_..._GSU1` appeared at **137.5 µs × 39.2 calls/step**. An `--enforce-eager` run agreed at 137.06 µs × 40 calls, so it is not a graph-capture artifact.
2. Call-chain attribution on the eager trace resolved it to `aten::mm < aten::matmul < aten::linear < vllm::rocm_unquantized_gemm < vllm::moe_forward_shared`.
3. Dispatch counts for the step were exactly **271 = 231 `wvSplitK` + 40 fallback**, with `LLMM1` never taken — i.e. exactly one linear per layer misses the ladder.
4. `shared_expert_gate.weight [1, 2048]` is the **only** 2-D weight in the model with `shape[0] == 1`, so the 40 fallbacks are fully accounted for.

## Proposed fix

A `[1, K]` weight is a dot product. Adding it as a third branch in the existing ladder:

```python
        elif m == 1:
            out = (x_view * weight.reshape(-1)).sum(-1, keepdim=True)
            if bias is not None:
                out = out + bias
            return out.reshape(*x.shape[:-1], weight.shape[0])
```

With that mounted over the container's copy, Part B's `m = 1` row goes **139.93 µs → 6.82 µs (20.5x)**, and every other row is unchanged within noise (7.88–8.08 µs). End-to-end on the model above: **+15.5% decode**, with the `Cijk` kernel gone from the profile entirely and `wvSplitK` untouched (17.58 vs 17.54 ms/step), i.e. nothing else moved.

The `bias is not None` arm is defensive only — transformers builds this layer with `bias=False`.

Relaxing the `LLMM1` branch to accept `m < 4` looks unattractive: it passes a hardcoded split factor of 4, and the `m % 4 == 0` guard appears to be a vectorization requirement. The patch deliberately fixes only `m == 1`; `m ∈ {2, 3, 5, 6, 7}` remain uncovered, and I have no model that exercises them, so I would rather not widen it speculatively. Happy to restructure however maintainers prefer.

## Numerics — please read before taking the patch as-is

**The output is not bitwise identical to `F.linear`, and the difference is not only summation order.**

`x_view * weight.reshape(-1)` is a bf16 × bf16 product, so **each of the K element products is rounded to bf16 before it is summed**. `F.linear` accumulates the products in fp32 without rounding the intermediates. `torch.sum` then accumulates those bf16 values in fp32, so the reduction step itself is fine — it is the elementwise product that loses precision the GEMM keeps.

Measured on random bf16 vectors at K=2048, max absolute difference was **3.1e-02 to 2.5e-01** across runs, on dot products of order ~45 — a small number of bf16 ulps, but a real and systematic difference rather than pure reordering.

If that matters, the safer formulation is to widen before multiplying:

```python
            out = (x_view.float() * weight.reshape(-1).float()).sum(-1, keepdim=True).to(x.dtype)
```

which should keep the fp32 intermediates the GEMM has. **I have not benchmarked that variant** — it adds two small casts to a 4 KiB operand, so I would expect it to stay far ahead of 140 µs, but I would not want that assumption in the patch without a maintainer confirming which trade-off is wanted. Padding `m` to 16 and slicing the result is a third option that would stay exact and keep it a GEMM.

For what it is worth on end-to-end behaviour with the bf16-product version applied: greedy decode matched the unpatched baseline for 794 of 1028 characters (~200 identical tokens) before diverging at a near-tie between two equally plausible continuations, and across 400 tool-calling benchmark problems there were **zero malformed tool calls**. That is evidence the patched model is healthy in absolute terms; it is **not** a controlled A/B, since I do not have an unpatched run of the same benchmark. A maintainer wanting a strict before/after should ask for one.

vLLM already selects different kernels by shape, so bitwise stability across shapes does not appear to be a property it currently guarantees — flagging it explicitly in case that changes which fix is preferred.

## Relationship to existing work on this call site

**#43187** proposes a Triton fusion for the *same* call site — collapsing the gate's skinny GEMM, sigmoid and broadcast-multiply into a single kernel, measured on MI355x. I believe the two are complementary rather than competing, because they attack different layers:

- #43187 removes **kernel-launch count and intermediate tensors** in `models/qwen2_moe.py`.
- This issue is that `layers/utils.py`'s dispatch ladder **never offers that GEMM a skinny path at all**, so it lands on a rocBLAS kernel measured at 17.7× the cost of the ones the ladder does select for neighbouring widths.

If #43187 lands, its fused kernel replaces this call site for the models it covers, and this fix still applies to every *other* `m == 1` linear reaching `rocm_unquantized_gemm_impl`. If it doesn't, this fix helps the call site on its own. Fixing the dispatch also makes the un-fused path a fairer baseline to measure a fusion against.

**#39280** (merged) added AITER Fused Shared Experts, which bypasses the call site at MoE-block level — but it is opt-in (`VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS`, default `False`) and requires AITER. **#37800**, which would have enabled it by default on ROCm, was closed. So on a default ROCm configuration, and on any configuration running with AITER off (as gfx1151 is here), the unfused path is what executes.

## Related

- #15830 introduced this dispatch (and, per its history, the `n <= 5` bound for spec decode).
- Point 2 under the reproducer may deserve a separate rocBLAS/hipBLASLt report: selecting a 128×32 tile with `GSU1` for a degenerate 1-row output is a poor heuristic regardless of what vLLM hands it.

### PR DESCRIPTION
**Pull request by tangzzycc**
## Purpose

Fixes #52631. ROCm currently sends unquantized GEMMs with `weight.shape[0] == 1` to the generic `F.linear` fallback because the `wvSplitK` branch only accepts `m > 8`. This shape is used by the Qwen shared-expert gate during batch-1 decode. The existing `wvSplitK` kernel already supports `M_in=1`, so this PR extends the Python dispatch condition and adds focused coverage.

This does not duplicate an existing PR. The issue author confirmed that no PR was in progress. 

## Test Plan

```bash
.venv/bin/python -m pytest tests/model_executor/layers/test_rocm_unquantized_gemm.py -v
.venv/bin/pre-commit run ruff-check --files vllm/model_executor/layers/utils.py tests/model_executor/layers/test_rocm_unquantized_gemm.py
.venv/bin/pre-commit run ruff-format --files vllm/model_executor/layers/utils.py tests/model_executor/layers/test_rocm_unquantized_gemm.py
```

The end-to-end A/B test used `Qwen/Qwen3.5-35B-A3B-GPTQ-Int4` on an AMD Radeon 8060S (`gfx1151`)

## Test Result

The focused test suite passed: `21 passed`.

| Metric | Before | After | Change |
| --- | ---: | ---: | ---: |
| BF16 `[1, 2048]` operator latency | 119.484 us | 3.161 us | 37.8x faster |
| Pure decode throughput | 23.74 tok/s | 26.77 tok/s | +12.79% |
| Decode latency per token | 42.125 ms | 37.350 ms | -11.34% |

The profiler confirmed the `wvSplitK` specialization with `YTILE=1`. All greedy output tokens were identical between the baseline and modified paths.

AI assistance was used during investigation, test development, and preparation of this PR. I reviewed every changed line and ran the tests and benchmarks above.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
</details>

### CODE REVIEW

**Review by big-yellow-duck — APPROVED** [review]
Reviewed the dispatch-only change and tests; no blocking concerns.

Independent reproduction on Radeon 8060S (gfx1151), ROCm 7.14 / torch 2.11:
- BF16 [1, 2048] stock dispatch: 137.9 us/call versus ~7.13 us for covered skinny widths.
- Direct m=1 wvSplitK selected wvSplitK_hf_sml_<__hip_bfloat16, 32, 1, 16, 8, 4, 1>, at 1.67 us GPU time and matched F.linear exactly for the sampled input.
- Full 40-layer Qwen3.5-MoE-compatible FP16 AWQ profile: slow rocBLAS launches dropped from 2,040 to 40 (the remaining 40 are prefill), while 2,000 decode calls moved to wvSplitK; profiled decode improved 43.629 to 40.948 ms/token.

The current pre-run-check failure is CI authorization/label gating, not a test failure. The existing kernel already handles M_in=1, so widening the Python guard is a focused, low-risk performance fix. Kernel redesign can remain separate.

**Comment by tangzzycc** [comment]
Hi @AndreasKaratzas @tjtanaa, thanks to @big-yellow-duck for the initial review. Could you also take a look at this PR when you have a chance? Thank you!

**Review by DarkLight1337 — APPROVED** [review]
